### ISSUE TITLE
Feature Request: make SoftMax application optional in runInference

### ISSUE THREAD

**Issue by avyan-k** [qupath/qupath-extension-wsinfer#69]
* How it works currently: a `applySoftmax` field is set to `true` in the `runInference` base method (see [here](https://github.com/qupath/qupath-extension-wsinfer/blob/dfab6802a88d9ba94d6f39531595da582e261f74/src/main/java/qupath/ext/wsinfer/WSInfer.java#L196)), which means the SoftMax function is always applied to the output of the model to generate class probabilities
* How it would work ideally: have the option to decide if SoftMax is applied, either through changing some of the method overloads or a global option to avoid disrupting existing functionality (see incoming PR for simplified implementation of the former option)
* Why I believe it is important/should be added: I would like to run inference using my own models, and they already apply softmax on their output. Since the SoftMax function is not idempotent, applying it a second time would "uniformise" the resulting class probabilities. To avoid this, I remove this last activation layer from my models, but this make the model architecture unclear and the TorchScript inconsistent from the base PyTorch model. I believe it is more consistent if `runInference` simply measures the output a model without modifying the output in any unexpected way, just as it does not modify the input beyond what is specified in the config.json file. Not applying the SoftMax automatically will also allow `runInference` to be used for other kinds of measurements beyond class probabilities, or perhaps other output activation functions to be applied instead of SoftMax.

**Comment by alanocallaghan** [comment]
Should this be an option or field in for example `WSInferModel` or `WSInferModelConfiguration`?

**Comment by avyan-k** [comment]
Yes, I agree this should be a field in `WSinferModel`. Ideally, it should be an option set in `WSInferModelConfiguration` as you mentioned, since it pulls directly from the config file that is based on [this schema](https://github.com/SBU-BMI/wsinfer-zoo/blob/main/wsinfer_zoo/schemas/model-config.schema.json), but I'm not sure how to do that without breaking compatibility with the original WSinfer pipeline while ensuring that `WSInferModelConfiguration` is still JSON serializable. Then I guess a boolean field in `WSinferModel` would work instead, which is false by default but then is overridden in `WSInferModelLocal` if needed. Another option would be to provide a custom `Translator` class from the DJL library as an argument to `runInference` to process the input and output in any way needed, including any custom transforms. [This class is already overridden to provide MPS support](https://github.com/qupath/qupath-extension-wsinfer/blob/main/src/main/java/qupath/ext/wsinfer/MpsSupport.java).

As a side note, I was wondering how debugging is meant to be done for writing QuPath extensions, as I'm unable to find any resources for this online. I currently just build the extension with gradlew, and test it directly with QuPath, but my assumption is that this is not the most efficient way of going about this.

**Comment by alanocallaghan** [comment]
I will have to tag @kaczmarj for thoughts on the JSON as this would also affect the python implementation.

> On a side note, I was wondering how debugging is meant to be done for writing QuPath extensions, as I'm unable to find any resources for this online. I currently just build the extension with gradlew, and test it directly with QuPath, but my assumption is that this is not the most efficient way of going about this.

The instructions here will make this much much easier :) 

https://github.com/qupath/qupath-extension-template?tab=readme-ov-file#run-qupath--the-extension

**Comment by kaczmarj** [comment]
i think this is a great idea. it might make most sense to put the softmax option in the json config (and to update the schema). 

we can add a new boolean property like `apply_softmax`, and make that optional, with the default being `true`.

**Comment by alanocallaghan** [comment]
Great! On the QuPath side it shouldn't take much work, so once you've updated the JSON spec and models I can update the extension

**Comment by avyan-k** [comment]
Hello, thank you for your suggestion about using the qupath extension template, it made building and testing the extension must more streamlined! I worked adding `apply_softmax` as an option to the model configuration in the softmax branch. For now, the `WSInferModelConfiguration` class has an `applySoftmax` attribute with a `SerializedName` annotation and default value of true, this value is overridden by the `apply_softmax` key in the `config.json` file when it is loaded for the model. Therefore, for all current models, SoftMax is applied by default. Once the JSON is updated in the [wsinfer repo](https://github.com/SBU-BMI/wsinfer), one update to be done before merging the branches would be to remove the default true value to enforce the fact that this option must be specified in the JSON for custom models. @kaczmarj I opened an issue on the wsinfer repo so that you or any other maintainer can work on it whenever they have some time, I would add a PR myself, but I do not know all the considerations that should be made for such a change.

About the PR for this repo, note that I squashed the previous commit, so it no longer exists on the branch. I did this to avoid showing edit history for files that should have never been edited (notably `WSInferController` for the UI), thus making the overall branch merge cleaner. Anyway, if the old changes I made were pulled, be sure to checkout off of that branch to avoid conflicts.
